### ISSUE TITLE
Trimming Error When Publishing Project Using Source Generator

### ISSUE THREAD

**Issue by ritchiecarroll** [dotnet/sdk#45253]
## Issue Summary

When publishing a project that uses a source generator with trimming enabled, the build fails with an error suggesting .NET Core 3.0+ is required, despite the project targeting .NET 9.0 and the source generator being marked as non-referenced build-time component.

## Version Used

Visual Studio 17.12.0 / .NET 9.0 / C# 13

## Steps to Reproduce

1. Create a .NET Standard 2.0 source generator project (.NET Standard 2.0 being required for analyzers)
2. Create a .NET 9.0 project that references this source generator, using code that depends on the generated output,
  generator referenced in the consuming project like so:
   ```xml
   <ProjectReference Include="..\MyGenerator.csproj" 
                     OutputItemType="Analyzer" 
                     ReferenceOutputAssembly="false" 
                     PrivateAssets="All" />
   ```
4. Build project normally with generated code
5. Attempt to _publish_ the consuming project, e.g., to "Folder", with trimming enabled by checking "Trim unused code" under "File publish options" in the "Folder" profile settings.

## Expected Behavior

The project should build / publish successfully with trimming enabled. Since the source generator is only used at build time and is marked with `ReferenceOutputAssembly="false"`, it should not be considered during the trimming process.

## Actual Behavior

Build fails with error `NETSDK1124`:
```
C:\Program Files\dotnet\sdk\9.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5):
Error NETSDK1124: Trimming assemblies requires .NET Core 3.0 or higher.
```

The error occurs despite:
1. The consuming project targeting .NET 9.0
2. The analyzer being marked as not referenced in the output
3. The analyzer being a build-time-only component

## Workaround

The issue can be worked around by conditionally by first building project normally, emitting generated files into a local folder. Folder needs to be excluded for normal builds, but available when publishing trimmed, then also excluding the source generator during trimmed publish. For example:

```xml
  <PropertyGroup>
    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
    <CompilerGeneratedFilesOutputPath>generated</CompilerGeneratedFilesOutputPath>
  </PropertyGroup>

  <ItemGroup Condition="'$(PublishTrimmed)' != 'true'">
    <Compile Remove="generated\**" />
    <EmbeddedResource Remove="generated\**" />
    <None Remove="generated\**" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\MyGenerator.csproj" 
      OutputItemType="Analyzer" 
      ReferenceOutputAssembly="false" 
      PrivateAssets="All" 
      Condition="'$(PublishTrimmed)' != 'true'" />
  </ItemGroup>
```

This works because the source generator has already generated the necessary files during the initial normal build phase, then they are available during the trimmed publish even without the generator reference.

**Comment by ritchiecarroll** [comment]
Same error message from command line:
```
> dotnet publish .\MyProject.csproj /p:PublishProfile=Properties\PublishProfiles\win-x64.pubxml

... lines omitted ...

  MyProject failed with 1 error(s) (0.0s)
    C:\Program Files\dotnet\sdk\9.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.CrossTargeting.targets(31,5): error NETSDK1129: The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, you must specify one of the following frameworks in order to publish: net9.0

Build failed with 1 error(s) in 1.9s
```

To me, this simply looks like trimmer is not skipping `ProjectReference` when `ReferenceOutputAssembly="false"`.

The workaround above seems to confirm this.

**Comment by ritchiecarroll** [comment]
For reference, one of the publish profiles used in testing error `win-x64.pubxml`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project>
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>Any CPU</Platform>
    <PublishDir>bin\Release\net9.0\publish\win-x64\</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <_TargetId>Folder</_TargetId>
    <TargetFramework>net9.0</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <PublishSingleFile>true</PublishSingleFile>
    <PublishTrimmed>true</PublishTrimmed>
  </PropertyGroup>
</Project>
```

**Comment by agocke** [comment]
Ah, I think this might be because of the way PublishProfiles work. I think a publish profile gets shoved into all dependent projects, not just the one you're building. @sbomer might have more info about that.

Two potential fixes.
 
1. Most preferred -- we suggest _not_ using publish profiles for trimming. We suggest putting `PublishTrimmed` directly into the project file. That will make sure that you see any potential trimming warnings on build as well as publish.
2. You could add a condition for TFM. If you change `<PublishTrimmed>true</PublishTrimmed>` to `<PublishTrimmed Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">true</PublishTrimmed>` I think that would prevent the property from kicking in to any pre-5.0 projects.

**Comment by agocke** [comment]
@vijayrkn I think we should remove the "trim" and "single file" checkboxes from the Publish Profile dialog. I've seen a few customers get tripped up by this now. Another issue is https://github.com/microsoft/devhome/issues/3946. The right checkbox that people want is actually this one in the "Publish Properties" page, not in the Publish Profiles dialog:
![Image](https://github.com/user-attachments/assets/415640fd-53ce-46e8-96fc-41ca16d5778b)

**Comment by agocke** [comment]
cc @eerhardt

**Comment by ritchiecarroll** [comment]
> Ah, I think this might be because of the way PublishProfiles work. I think a publish profile gets shoved into all dependent projects, not just the one you're building. [@sbomer](https://github.com/sbomer) might have more info about that.
> 
> Two potential fixes.
> 
> 1. Most preferred -- we suggest _not_ using publish profiles for trimming. We suggest putting `PublishTrimmed` directly into the project file. That will make sure that you see any potential trimming warnings on build as well as publish.
> 2. You could add a condition for TFM. If you change `<PublishTrimmed>true</PublishTrimmed>` to `<PublishTrimmed Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">true</PublishTrimmed>` I think that would prevent the property from kicking in to any pre-5.0 projects.

@agocke  - good news, your suggested option 2 works!

Here's an operational publish profile that includes TFM conditional:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project>
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>Any CPU</Platform>
    <PublishDir>bin\Release\net9.0\publish\win-x64\</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <_TargetId>Folder</_TargetId>
    <TargetFramework>net9.0</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <PublishSingleFile>true</PublishSingleFile>
    <PublishTrimmed Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">true</PublishTrimmed>
  </PropertyGroup>
</Project>
```

One side effect (perhaps excepted), the UI option for this now looks _unchecked_:
![Image](https://github.com/user-attachments/assets/18385023-adc3-4eb5-8de2-24af6df9466a)

**Comment by ritchiecarroll** [comment]
@agocke  **_UPDATE:_** Got option 1 to work if I _remove_ `<PublishedTrimmed>` option from the `.pubxml` file altogether and then add `<PublishedTrimmed>true</PublishedTrimmed>` to my main project file. 

However, as you mention, if I accidentally check the box for "Trim unused code" in the "Profile settings > File publish options", the error will return... For that matter, option 2 may have the same issue, with error returning, if checking box (which _appears_ unchecked) removes conditional TFM.

Anyway, hopefully this detail helps you all produce a good resolution for an update to VS or SDK, etc.

Regardless, in the mean time it is good to have some simpler work around options for now, per your recommendations. Thanks!

**Comment by agocke** [comment]
Yup, sorry, the suggestion was to _remove_ `<PublishTrimmed>` from the pubxml and put it into the project file -- basically avoid the pubxml entirely.

And yeah, I think this check box is just a bad idea. As I screenshotted above, there's also a checkbox in the Properties page. That's the good one. We should delete the one in the publish profile.
